cpp/src/join/detail/traversal.cuh

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/gather.h>
+#include <thrust/scan.h>
+#include <thrust/scatter.h>
+#include <thrust/tuple.h>
+
+#include <tuple>
+
+#include "indexing/construction/detail/utilities.cuh"
+#include "utility/z_order.cuh"
+
+namespace cuspatial {
+namespace detail {
+
+template <typename LengthsIter, typename OffsetsIter>
+inline std::tuple<cudf::size_type,
+                  rmm::device_uvector<uint8_t>,
+                  rmm::device_uvector<uint8_t>,
+                  rmm::device_uvector<uint32_t>,
+                  rmm::device_uvector<uint32_t>>
+descend_quadtree(LengthsIter lengths,
+                 OffsetsIter offsets,
+                 cudf::size_type num_quads,
+                 rmm::device_uvector<uint8_t> &quad_types,
+                 rmm::device_uvector<uint8_t> &quad_levels,
+                 rmm::device_uvector<uint32_t> &quad_node_indices,
+                 rmm::device_uvector<uint32_t> &quad_poly_indices,
+                 cudaStream_t stream)
+{
+  // compute the total number of child nodes
+  auto num_children =
+    thrust::reduce(rmm::exec_policy(stream)->on(stream), lengths, lengths + num_quads);
+
+  // exclusive scan on the number of child nodes to compute the offsets
+  rmm::device_uvector<uint32_t> parent_offsets(num_quads, stream);
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream)->on(stream), lengths, lengths + num_quads, parent_offsets.begin());
```

Comment:
You could combine these. No need to compute the num_children with a separate reduction if you do an inclusive scan  offset by 1 and read back the last element:

(note two `+ 1` offsets below, and also need to set the first element of the scan results to 0)

```suggestion
  // scan on the number of child nodes to compute the offsets
  rmm::device_uvector<uint32_t> parent_offsets(num_quads + 1, stream);
  thrust::inclusive_scan(
    rmm::exec_policy(stream)->on(stream), lengths, lengths + num_quads, parent_offsets.begin() + 1);
  parent_offsets.set_element_async(0, 0, stream);
  auto num_children = parent_offsets.back_element(stream); // synchronizes stream
```

I think the synchronization here is less heavy handed than `thrust::reduce` as well (and doesn't happen until you need `num_children`.

Reply:
Great suggestion! I updated the implementation.